internal/graph/check.go

```diff
@@ -459,7 +460,9 @@ func (c *LocalChecker) ResolveCheck(
 		return nil, fmt.Errorf("relation '%s' undefined for object type '%s'", relation, objectType)
 	}
 
-	resp, err := union(ctx, c.concurrencyLimit, c.checkRewrite(ctx, req, rel.GetRewrite()))
+	reqDatastore := storagewrappers.NewRequestTupleReader(c.ds, req.GetContextualTuples())
```

Comment:
I could also make this object a part of `ResolveCheckRequest` 🤔

Reply:
I'm not sure I'm aligned with this change and/or the one you mentioned as an alternative (in the ResolveCheckRequest). The reason we introduced the `CombinedTupleReader` in the first place to was move this responsibility outside of Check.

We moved datastore specific code outside of Check in the first place so as to not conflate the implementation details with contextual tuples. As far as Check is concerned, it simply yields tuples from an iterator that comes from the database iterator provided. Those tuples may be due to contextual tuples or ones that have been persisted, but Check doesn't care. Tuples are tuples.

It's not Checks responsibility to construct a datastore derivative from the contextual tuples provided in the request, that's the upstream callers responsibility if that's the intent.

Moving this inside of Check is tangential to that initial intent.


So I'm a little torn on these changes..

Reply:
>  As far as Check is concerned, it simply yields tuples from an iterator that comes from the database iterator provided. Those tuples may be due to contextual tuples or ones that have been persisted, but Check doesn't care. Tuples are tuples. 

That specific abstraction is provided by the `requestTupleReader`

> It's not Checks responsibility to construct a datastore derivative from the contextual tuples provided in the request, that's the upstream callers responsibility if that's the intent.

I don't agree. Check has a field in the request called `ContextualTuples`. It should know how to handle that.

Look at the test code we have today:

```go
checker := NewLocalChecker(storagewrappers.NewCombinedTupleReader(ds, test.contextualTuples))

res, err := checker.ResolveCheck(ctx, &ResolveCheckRequest{
	ContextualTuples:   test.contextualTuples,
	// ...
})
```

As a developer writing this, I ask myself: why am i given two places to pass in the contextual tuples? What happens if I forget to pass in one of them?

This is the ambiguity I want to get rid of.

Reply:
I redact my original comment. It's not unreasonable to expect the Check implementation to create a derivative iterator that yields the contextual tuples.  The original goal was to avoid manually iterating over the contextual tuples and validating/filtering them and doing evaluations against them etc.. We shouldn't have to treat the contextual tuples any different than tuples coming from the storage iterator. As long as we keep contextual tuples abstracted away from the primary Check code paths, then I'm happy 😄 

Since the contextual tuples are request specific and the derivative storage implementation has to be constructed per request, then I'm ok with this approach. But if we start seeing very specific contextual tuple code in critical code paths, then we should reconsider what we're doing.